### PR TITLE
IAM Policy Assignment: S2S Policy Assignments ET and AG

### PR DESCRIPTION
**Pull request by siddhuvarma1997**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates OR Closes #0000

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccXXX'

...
```

### CODE REVIEW

**Comment by siddhuvarma1997** [comment]
image

**Comment by siddhuvarma1997** [comment]
image

**Comment by siddhuvarma1997** [comment]
terraform config:

`resource "ibm_iam_policy_template" "policy_s2stemplate" {
	name = "TestTerraformS2SAccount"
	policy {
		type = "authorization"
		description = "Test terraform enterprise S2S"
		resource {
			attributes {
				key = "serviceName"
				operator = "stringEquals"
				value = "kms"
			}
		}
		subject {
			attributes {
				key = "serviceName"
				operator = "stringEquals"
				value = "compliance"
			}
		}
		roles = ["Reader"]
	}
	committed=true
}
resource "ibm_iam_policy_assignment" "policy_assignment" {
	version = "1.0"
	target  ={
		type = "Enterprise"
		id = "562c2a1307534d648e9865f421e454d0"
	}
	templates{
		id = ibm_iam_policy_template.policy_s2stemplate.template_id 
		version = 1
	}
	template_version="2"

}


	   resource "ibm_iam_policy_template_version" "template_version" {
	template_id = ibm_iam_policy_template.policy_s2stemplate.template_id 
	policy {
		type = "authorization"
		description = "Test terraform enterprise S2S"
		subject {
								attributes {
										key = "serviceName"
										operator = "stringEquals"
										value = "compliance"
								}
						}
		resource {
			attributes {
				key = "serviceName"
				operator = "stringEquals"
				value = "appid"
			}
		}
		roles = ["Reader"]
	}
       committed=true
}
`

`terraform show
# ibm_iam_policy_assignment.policy_assignment:
resource "ibm_iam_policy_assignment" "policy_assignment" {
    accept_language     = "default"
    account_id          = "e3aa0adff348470f803d4b6e53d625cf"
    created_at          = "2024-09-12T06:21:48.586Z"
    created_by_id       = "IBMid-696000BLE3"
    href                = "https://iam.test.cloud.ibm.com/v1/policy_assignments/policyAssignment-f19d12fd-0b7b-4ec5-bedf-1a9b74aa4733"
    id                  = "policyAssignment-f19d12fd-0b7b-4ec5-bedf-1a9b74aa4733"
    last_modified_at    = "2024-09-12T06:25:34.610Z"
    last_modified_by_id = "IBMid-696000BLE3"
    resources           = [
        {
            policy = [
                {
                    error_message    = []
                    resource_created = [
                        {
                            id = "6aca2e6b-01c5-45a2-b565-10d90023fbd2"
                        },
                    ]
                    status           = null
                },
            ]
            target = {
                "id"   = "00a430d3c66c472491cfc863598f123a"
                "type" = "Account"
            }
        },
        {
            policy = [
                {
                    error_message    = []
                    resource_created = [
                        {
                            id = "a62f9592-3097-4afd-85ea-0f5ef644aa0f"
                        },
                    ]
                    status           = null
                },
            ]
            target = {
                "id"   = "002f345a049b4f11ae6206661e5cb438"
                "type" = "Account"
            }
        },
        {
            policy = [
                {
                    error_message    = []
                    resource_created = [
                        {
                            id = "69e8430e-2556-4b07-bb56-c7e34744e31d"
                        },
                    ]
                    status           = null
                },
            ]
            target = {
                "id"   = "013a9d8f74954953948aebcbb3e859f1"
                "type" = "Account"
            }
        },
        {
            policy = [
                {
                    error_message    = []
                    resource_created = [
                        {
                            id = "932b6435-f9e0-4d32-a053-767af1b75acf"
                        },
                    ]
                    status           = null
                },
            ]
            target = {
                "id"   = "0ddc8e9eb0894383b48791ff4ba0b80e"
                "type" = "Account"
            }
        },
        {
            policy = [
                {
                    error_message    = []
                    resource_created = [
                        {
                            id = "133f92e4-de33-4a22-9663-31b2d8a2871a"
                        },
                    ]
                    status           = null
                },
            ]
            target = {
                "id"   = "08e31fa9e7df40ca9ce56685923aa756"
                "type" = "Account"
            }
        },
        {
            policy = [
                {
                    error_message    = []
                    resource_created = [
                        {
                            id = "6e036713-98c9-429a-99b4-8b93a5eb758c"
                        },
                    ]
                    status           = null
                },
            ]
            target = {
                "id"   = "028b7826db48435b9e007e1a85fff8ee"
                "type" = "Account"
            }
        },
        {
            policy = [
                {
                    error_message    = []
                    resource_created = [
                        {
                            id = "706e7173-6ce5-4a8a-8ad1-d7af896e9ac4"
                        },
                    ]
                    status           = null
                },
            ]
            target = {
                "id"   = "17db07eebd3548e5a4975356c2a9da77"
                "type" = "Account"
            }
        },
        {
            policy = [
                {
                    error_message    = []
                    resource_created = [
                        {
                            id = "329554e9-604e-45ed-ac44-7e8e5683b2cd"
                        },
                    ]
                    status           = null
                },
            ]
            target = {
                "id"   = "2ac280ab35024a9185bf628af9219835"
                "type" = "Account"
            }
        },
        {
            policy = [
                {
                    error_message    = []
                    resource_created = [
                        {
                            id = "d08285e3-6bd8-4ea9-95b0-6e329b8fcdb5"
                        },
                    ]
                    status           = null
                },
            ]
            target = {
                "id"   = "55ada8a4a75a48f58df782ba31c8e2ca"
                "type" = "Account"
            }
        },
        {
            policy = [
                {
                    error_message    = []
                    resource_created = [
                        {
                            id = "f6423a3c-c5aa-49a0-833b-1b42026c72c5"
                        },
                    ]
                    status           = null
                },
            ]
            target = {
                "id"   = "5dd10bd7e9a44ac990813d234649a752"
                "type" = "Account"
            }
        },
        {
            policy = [
                {
                    error_message    = []
                    resource_created = [
                        {
                            id = "b2712baf-a970-4fdf-99c5-037632186458"
                        },
                    ]
                    status           = null
                },
            ]
            target = {
                "id"   = "5b580a95e71c4afb99fc601d8e66e281"
                "type" = "Account"
            }
        },
        {
            policy = [
                {
                    error_message    = []
                    resource_created = [
                        {
                            id = "d75d0c41-2182-4b9a-9836-bea20aa06c64"
                        },
                    ]
                    status           = null
                },
            ]
            target = {
                "id"   = "6e42a47f8ae143fa9accc55dfe22096f"
                "type" = "Account"
            }
        },
        {
            policy = [
                {
                    error_message    = []
                    resource_created = [
                        {
                            id = "01d55701-0397-4f32-8d5f-4d8ea9fe8741"
                        },
                    ]
                    status           = null
                },
            ]
            target = {
                "id"   = "8c2f25994fb74fe18539205580885559"
                "type" = "Account"
            }
        },
        {
            policy = [
                {
                    error_message    = []
                    resource_created = [
                        {
                            id = "67bbb547-9b8b-4e69-a4fb-15ca4d3c971a"
                        },
                    ]
                    status           = null
                },
            ]
            target = {
                "id"   = "8c9c5dbf7d564e068cead0e327776451"
                "type" = "Account"
            }
        },
        {
            policy = [
                {
                    error_message    = []
                    resource_created = [
                        {
                            id = "9737afa2-433d-4572-9a9f-0aae89f73ddd"
                        },
                    ]
                    status           = null
                },
            ]
            target = {
                "id"   = "9699deb12e3d4430b1c94cd0e07a4699"
                "type" = "Account"
            }
        },
        {
            policy = [
                {
                    error_message    = []
                    resource_created = [
                        {
                            id = "88fe86c9-5fcd-4293-adde-86da86bb1077"
                        },
                    ]
                    status           = null
                },
            ]
            target = {
                "id"   = "948ee4a53bfd435f8b8c195e08f2bbac"
                "type" = "Account"
            }
        },
        {
            policy = [
                {
                    error_message    = []
                    resource_created = [
                        {
                            id = "cfb9cbc2-7220-4a3d-ad25-84bfbd989092"
                        },
                    ]
                    status           = null
                },
            ]
            target = {
                "id"   = "9e6c355206964be9a17d41b90fe77844"
                "type" = "Account"
            }
        },
        {
            policy = [
                {
                    error_message    = []
                    resource_created = [
                        {
                            id = "b6653b0f-22ed-4471-9a3e-1adebe1fea50"
                        },
                    ]
                    status           = null
                },
            ]
            target = {
                "id"   = "ecaade7fe62a4891adb701565d235c74"
                "type" = "Account"
            }
        },
        {
            policy = [
                {
                    error_message    = []
                    resource_created = [
                        {
                            id = "f1102b3c-b1b1-42cd-bb85-3e8b98fe3668"
                        },
                    ]
                    status           = null
                },
            ]
            target = {
                "id"   = "c169ec2b63fe472bba59b7f39496b047"
                "type" = "Account"
            }
        },
        {
            policy = [
                {
                    error_message    = []
                    resource_created = [
                        {
                            id = "1b8b9b27-e8a7-42d6-a0ee-b891b631938b"
                        },
                    ]
                    status           = null
                },
            ]
            target = {
                "id"   = "fe1839a1b56a4f4fa12239825fa5f4d8"
                "type" = "Account"
            }
        },
    ]
    status              = "succeeded"
    target              = {
        "id"   = "562c2a1307534d648e9865f421e454d0"
        "type" = "Enterprise"
    }
    template            = {
        "id"      = "policyTemplate-8db25168-454b-4928-b68b-60c83dbd45e8"
        "version" = "2"
    }
    template_version    = "2"
    version             = "1.0"

    templates {
        id      = "policyTemplate-8db25168-454b-4928-b68b-60c83dbd45e8"
        version = "1"
    }
}

# ibm_iam_policy_template.policy_s2stemplate:
resource "ibm_iam_policy_template" "policy_s2stemplate" {
    account_id  = "e3aa0adff348470f803d4b6e53d625cf"
    committed   = true
    id          = "policyTemplate-8db25168-454b-4928-b68b-60c83dbd45e8/1"
    name        = "TestTerraformS2SAccount"
    template_id = "policyTemplate-8db25168-454b-4928-b68b-60c83dbd45e8"
    version     = "1"

    policy {
        description   = "Test terraform enterprise S2S"
        pattern       = null
        roles         = [
            "Reader",
        ]
        rule_operator = null
        type          = "authorization"

        resource {
            attributes {
                key      = "serviceName"
                operator = "stringEquals"
                value    = "kms"
            }
        }

        subject {
            attributes {
                key      = "serviceName"
                operator = "stringEquals"
                value    = "compliance"
            }
        }
    }
}

# ibm_iam_policy_template_version.template_version:
resource "ibm_iam_policy_template_version" "template_version" {
    account_id  = "e3aa0adff348470f803d4b6e53d625cf"
    committed   = true
    id          = "policyTemplate-8db25168-454b-4928-b68b-60c83dbd45e8/2"
    name        = "TestTerraformS2SAccount"
    template_id = "policyTemplate-8db25168-454b-4928-b68b-60c83dbd45e8"
    version     = "2"

    policy {
        description   = "Test terraform enterprise S2S"
        pattern       = null
        roles         = [
            "Reader",
        ]
        rule_operator = null
        type          = "authorization"

        resource {
            attributes {
                key      = "serviceName"
                operator = "stringEquals"
                value    = "appid"
            }
        }

        subject {
            attributes {
                key      = "serviceName"
                operator = "stringEquals"
                value    = "compliance"
            }
        }
    }
}
`

**Comment by hkantare** [comment]
update in docs the timeout section

File: ibm/service/iampolicy/resource_ibm_iam_policy_assignment.go [hunk 1]
```diff
@@ -477,3 +532,114 @@ func ResourceIBMPolicyAssignmentAssignmentTemplateDetailsToMap(model *iampolicym
 	}
 	return modelMap, nil
 }
+
+func waitForAssignment(timeout time.Duration, meta interface{}, d *schema.ResourceData, refreshFn func(string, interface{}) resource.StateRefreshFunc) (interface{}, error) {
+
+	stateConf := &resource.StateChangeConf{
+		Pending:      []string{InProgress},
```
[the remark this answers is not in the document]
Comment (siddhuvarma1997) [reply ibm/service/iampolicy/resource_ibm_iam_policy_assignment.go:539]: Hi @hkantare we have only in progress and next the record will be deleted there is no 'deleting' state

File: ibm/service/iampolicy/resource_ibm_iam_policy_assignment.go [hunk 2]
```diff
@@ -433,11 +465,34 @@ func resourceIBMPolicyAssignmentDelete(context context.Context, d *schema.Resour
...
-		tfErr := flex.TerraformErrorf(err, fmt.Sprintf("DeletePolicyAssignmentWithContext failed: %s", err.Error()), "ibm_policy_assignment", "delete")
-		log.Printf("[DEBUG]\n%s", tfErr.GetDebugMessage())
-		return tfErr.GetDiag()
+		return diag.FromErr(err)
+	}
+
+	response, err := iamPolicyManagementClient.DeletePolicyAssignmentWithContext(context, deletePolicyAssignmentOptions)
+	if err != nil {
+		if response != nil && response.StatusCode == 404 {
+			log.Printf("[DEBUG] Policy assignment not found during delete, assuming it's already deleted: %s", d.Id())
```
[the remark this answers is not in the document]
Comment (siddhuvarma1997) [reply ibm/service/iampolicy/resource_ibm_iam_policy_assignment.go:477]: Yes @hkantare we can return instead of logging the error I will add return

**Comment by siddhuvarma1997** [comment]
image

**Comment by hkantare** [comment]
https://github.com/IBM-Cloud/terraform-provider-ibm/pull/5624/files#r1756821671

This comment is n't addressed we are having same wait for creation/deletion

```


func waitForAssignment(timeout time.Duration, meta interface{}, d *schema.ResourceData, refreshFn func(string, interface{}) resource.StateRefreshFunc) (interface{}, error) {

	stateConf := &resource.StateChangeConf{
		Pending:      []string{InProgress},
		Target:       []string{complete},
		Refresh:      refreshFn(d.Id(), meta),
		Delay:        30 * time.Second,
		PollInterval: time.Minute,
		Timeout:      timeout,
	}

	return stateConf.WaitForState()
}

```

The target state for delete also be "completed"?
we see only failed and InProgress sent as part of isAccessPolicyAssignedDeleted method

**Comment by siddhuvarma1997** [comment]
Hi @hkantare for delete we have only `in progress` and next the record will be removed either failure there is no other state

**Review by hkantare — APPROVED** [review]

**Pull request merged**